clarity/contracts/arkadiko-claim-yield-v1-1.clar

```diff
@@ -0,0 +1,178 @@
+;; However, this is not a feasible solution since it does not scale in number of vaults
+;; N vaults would require N transactions
+;; Therefore, we have this yield claimer that allows a user to claim their yield each cycle (or few cycles)
+;; We might deprecate this claimer once Stacks throughput improves or we find a solution that scales better
+;;
+
+(use-trait vault-trait .arkadiko-vault-trait-v1.vault-trait)
+(use-trait ft-trait .sip-010-trait-ft-standard.sip-010-trait)
+(use-trait collateral-types-trait .arkadiko-collateral-types-trait-v1.collateral-types-trait)
+
+(define-constant ERR-NOT-AUTHORIZED u40401)
+(define-constant ERR-NOTHING-TO-CLAIM u40402)
+
+(define-map claims
+  { vault-id: uint }
+  {
+    ustx: uint
+  }
+)
+
+(define-read-only (get-claim-by-vault-id (vault-id uint))
+  (default-to
+    {
+      ustx: u0
+    }
+    (map-get? claims { vault-id: vault-id })
+  )
+)
+
+(define-public (add-claims (recipients (list 200 { to: uint, ustx: uint })))
+  (begin
+    (asserts! (is-eq tx-sender (contract-call? .arkadiko-dao get-dao-owner)) (err ERR-NOT-AUTHORIZED))
+
+    (map add-claim recipients)
+    (ok true)
+  )
+)
+
+(define-public (add-claim (recipient { to: uint, ustx: uint }))
+  (let (
+    (claim-entry (get-claim-by-vault-id (get to recipient)))
+  )
+    (asserts! (is-eq tx-sender (contract-call? .arkadiko-dao get-dao-owner)) (err ERR-NOT-AUTHORIZED))
+
+    (try! (stx-transfer? (get ustx recipient) tx-sender (as-contract tx-sender)))
+    (map-set claims { vault-id: (get to recipient) } { ustx: (+ (get ustx recipient) (get ustx claim-entry)) })
+    (ok true)
+  )
+)
+
+;; @desc Claim PoX yield and add as collateral
+;; @param vault-id; your vault ID
+;; @param reserve; active STX reserve
+;; @param coll-type; active collateral types contract
+;; @post boolean; returns true if claim was succesful
+(define-public (claim
+  (vault-id uint)
+  (reserve <vault-trait>)
+  (coll-type <collateral-types-trait>)
+  (stack-yield bool)
+)
+  (let (
+    (vault (contract-call? .arkadiko-vault-data-v1-1 get-vault-by-id vault-id))
+    (stacker-name (get stacker-name vault))
+    (claim-entry (get-claim-by-vault-id vault-id))
+    (sender tx-sender)
+  )
+    (asserts! (is-eq tx-sender (get owner vault)) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of reserve) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "stx-reserve"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of coll-type) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "collateral-types"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (> (get ustx claim-entry) u0) (err ERR-NOTHING-TO-CLAIM))
+
+    (try! (as-contract (stx-transfer? (get ustx claim-entry) tx-sender sender)))
+    (try! (contract-call? .arkadiko-freddie-v1-1 deposit vault-id (get ustx claim-entry) reserve .arkadiko-token coll-type))
+
+    (if (is-eq stack-yield true)
```

Comment:
@nieldeckx this will always fail unless the user does this when the stacker is in a cooldown cycle (i.e. after the 3 PoX cycles), but we pay out per PoX cycle.

Why? Check the `enable-vault-withdrawals` method in a stacker. We have the following assertion there:
```
    (asserts!
      (or
        (is-eq u0 (var-get stacking-stx-stacked))
        (>= burn-block-height (var-get stacking-unlock-burn-height))
      )
      (err ERR-BURN-HEIGHT-NOT-REACHED)
    )
```

Amount of STX stacked is not 0 and burn block height is not greater than the unlock burn height yet (since we stack for 3 PoX cycles). As a result, it will always throw `(err ERR-BURN-HEIGHT-NOT-REACHED)`.

We can keep this code, but it will only be useful for people claiming when we are in a cooldown.

Reply:
Now that I look at the stacker, I think maybe we didn't even need that assertion, since a user's tokens are stacked anyway (they can't be withdrawn/unlocked), so it would be safe the let the user unlock the vault.

Too late for now either way.

Reply:
@philipdesmedt got it, thx for the info.